docs/design/gh-ost-integration.md

```diff
@@ -0,0 +1,474 @@
+
+```Go
+const TaskCheckDatabaseSchemaUpdateGhostCutover = "bb.task-check.database.schema.update.ghost.cutover"
+type TaskCheckDatabaseSchemaUpdateGhostCutoverPayload struct {
+}
+```
+
+### Crash recovery
+
+If it crashes halfway when migrating on `tablename`, we will end up with the original `tablename`, and `_tablename_gho`, `_tablename_ghc`. We just drop `_tablename_gho` and `_tablename_ghc`, rerun gh-ost.
+
+However, we need to refactor pre-migration validation and migration status, because the current logic doesn't allow migration to have the same version, rerun will lead to failure. We can refactor to reuse version to keep it idempotent.
+
+Another approach is to abandon the previous version and start a new issue with a new version. The benefit is that we don't have to support rerun, so it's less complex.
+
+For the first iteration, we just let it fail.
+
+### After Migration
+
+After migration is finished, we will have `tablename` and `_tablename_del`, the latter is the original table. We are too scared to delete the original table for the user, so we'll generate a task to delete the table if approved by user.
+
+### Task dependency
+
+Tasks under the same stage run one by one by now. With gh-ost integration, we introduce task dependency. As mentioned above, the "cut-over" task depends on the "run" task. A task can only be running if all of its dependencies are done. Tasks that are not blocked by dependencies can run in parallel.
+
+To store the dependency relationship, we can add a field in Task.
+
+```Go
+type Task struct {
+    // BlockedBy stores TaskIDs
+    // Task cannot be running if not all of the BlockedBy tasks are done
+    BlockedBy []int
+}
+```
+
+However, we can only know the `TaskID` of a `Task` after insertion, so we have to update `Task` after all tasks in the same stage are inserted into the database.
+
+### MigrationContext
+
+We will introduce a new issue type for gh-ost. This is how we tell the difference between gh-ost and direct DDL.
+
+And a new task payload type and task executor for each execution task.
+
+```Go
+// issue related
+
+const IssueDatabaseSchemaUpdateGhost IssueType = "bb.issue.database.schema.update.ghost"
+
+type UpdateSchemaGhostDetail struct {
+    DatabaseID int `json:"databaseId"`
+    DatabaseName string `json:"databaseName"`
+    Statement string `json:"statement"`
+    EarliestAllowedTs int64 `jsonapi:"attr,earliestAllowedTs"`
+}
+
+type UpdateSchemaGhostContext struct {
+    MigrationType db.MigrationType `json:"migrationType"`
+    UpdateSchemaDetailList []*UpdateSchemaGhostDetail `json:"updateSchemaDetailList`
+    VCSPushEvent *vcs.PushEvent
+}
+
+// task related
+
+const TaskDatabaseSchemaUpdateGhost TaskType = "bb.task.database.schema.update.ghost"
```

Comment:
To be consistent with other two types of gh-ost types, we can name it "bb.task.database.schema.update.ghost.sync" for syncing data?

Reply:
Done.